controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+	// desired spec since that might involve moving the VM across
+	// compute and storage boundaries.
+	// VirtualMachines in deleted state are not excluded.
+	filteredVMs := make([]*vmopv1.VirtualMachine, 0, len(allVMs.Items))
+	for idx := range allVMs.Items {
+		vm := &allVMs.Items[idx]
+		// Skip this VirtualMachine if it is controlled by someone else.
+		if shouldExcludeVirtualMachine(ctx.ReplicaSet, vm) {
+			continue
+		}
+
+		// Attempt to adopt VirtualMachine if it has no controller references.
+		if metav1.GetControllerOf(vm) == nil {
+			if err := r.adoptOrphan(ctx, ctx.ReplicaSet, vm); err != nil {
+				ctx.Logger.Error(err, "VirtualMachineReplicaSet failed to adopt VirtualMachine", "vm", vm.Name)
+				r.Recorder.Warnf(ctx.ReplicaSet, "FailedAdopt", "Failed to adopt VirtualMachine %q: %v", vm.Name, err)
+
+				continue
+			}
+			ctx.Logger.Info("Adopted VirtualMachine", "vm", vm.Name)
+			r.Recorder.Eventf(ctx.ReplicaSet, "SuccessfulAdopt", "Adopted VirtualMachine %q", vm.Name)
+		}
+
+		filteredVMs = append(filteredVMs, vm)
+	}
+
+	// If not already present, add a label specifying the VirtualMachineReplicaSet name to VirtualMachines.
+	for _, vm := range filteredVMs {
+		// Note: MustEqualValue is used here as the value of this label will be
+		// a hash if the VirtualMachineReplicaSet name is longer than 63 characters.
+		if rsName, ok := vm.Labels[vmopv1.VirtualMachineReplicaSetNameLabel]; ok &&
+			MustEqualValue(ctx.ReplicaSet.Name, rsName) {
+
+			continue
+		}
+
+		res, err := controllerutil.CreateOrPatch(ctx, r.Client, vm, func() error {
+			// MustFormatValue is used here as the value of this label will be a
+			// hash if the VirtualMachineReplicaSet name is longer than max allowed
+			// label value length of 63 characters.
+			vm.Labels[vmopv1.VirtualMachineReplicaSetNameLabel] = util.MustFormatValue(ctx.ReplicaSet.Name)
+
+			return nil
+		})
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to apply %s label to VirtualMachine %q",
+				vmopv1.VirtualMachineReplicaSetNameLabel, vm.Name)
+		}
+
+		ctx.Logger.V(5).Info(
+			"Updated VirtualMachine with the VirtualMachineReplicaSet name label",
+			"result", res,
+			"vm", vm,
+			"key", vmopv1.VirtualMachineReplicaSetNameLabel,
+			"value", util.MustFormatValue(ctx.ReplicaSet.Name))
+
+		// TODO: Propagate the Deployment label from Deployment to replica set to VirtualMachine if
+		// it is set on the replica set.
+	}
+
+	// TODO: Handle in place propagation of fields to machines before syncing replicas.
+
+	syncErr := r.syncReplicas(ctx, ctx.ReplicaSet, filteredVMs)
+
+	// Update the status of the VirtualMachineReplicaSet even in case of error
+	// since syncing might have resulted in replicas being added or removed.
+	r.updateStatus(ctx, ctx.ReplicaSet, filteredVMs)
+
+	if syncErr != nil {
+		return ctrl.Result{}, errors.Wrapf(syncErr, "failed to sync VirtualMachineReplicaSet replicas")
```

Comment:
We are not using `github.com/pkg/errors` anymore. Please change this to:

```go
fmt.Errorf("failed to sync VirtualMachineReplicaSet replicas: %w", err)
```

Please see https://github.com/vmware-tanzu/vm-operator/pull/538 for more information.